### ISSUE TITLE
Activate rules only for specific configurations or build types

### ISSUE THREAD

**Issue by AliSoftware** [realm/SwiftLint#799]
As stated in #727 there can be a interest in activating some rules ONLY in specific configurations (Debug vs Release) and/or ONLY if doing an "Analyze" (Cmd-Maj-B) but not a standard "Build" (Cmd-B).

The most interesting of these is probably to only activate rules on Analyze, this way you could still error for force-unwraps, line length or stuff like that in standard Builds, and only warn about cyclomatic complexity and TODOs in Analyze

I guess today one could imagine having a dedicated `.swiftlint-analyze.yml` file and configure its Build Phase to use `.swiftlint.yml` on standard Builds and `.swiftlint-analyze.yml` on Analyze builds, but maybe it would be even cooler if that feature was integrated in SwiftLint directly, like as an additional parameter for each rule or something?

**Comment by scottrhoyt** [comment]
I suppose one way to implement this would be to create new a new severity level called analyze. During normal builds, you could filter out rules with this severity (for rules with multiple severity levels, you could remove this level). However, doing so you would still need to add a flag to `swiftlint` to let it know how to process analyze because I don’t know of any way to determine the type of build being done otherwise. So you would still need different build scripts for a normal vs. analyze build. Because of that, I think the complexity vs. utility of this is debatable because all you gain is a single configuration file.

**Comment by AliSoftware** [comment]
I see your point.

New idea then: ability to easily switch to "Opt-In only" mode in configuration file, to _only_ list the rules we want to enable explicitly, without having to disable all the default ones one by one except the few ones we want.

It would then be way easier to create configurations where you want some rules to be executed for the Build action and others to be executed for Analyze action, by having the `.swiftlint.yml` config file for Build only disable `todo` and the `.swiftlint-analyze.yml` config file used for Analyze to only enable `todo`

---

```sh
if [ "${RUN_CLANG_STATIC_ANALYZER}" == "YES" ]; then
  swiftlint --config ".swiftlint-analyze.yml"
else
  swiftlint
fi
```

```sh
$ cat .swiftlint.yml
disabled_rules:
 - todo
$ cat .swiftlint-analyze.yml
enabled_rules:
 - todo
```

**Comment by marcelofabri** [comment]
@AliSoftware We have `whitelist_rules` which could be what you've described on your last comment. Can we close this?

**Comment by AliSoftware** [comment]
Indeed 👍
